akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -257,3 +233,89 @@ private[akka] final class HeadOptionStage[T] extends GraphStageWithMaterializedV
     }, p.future)
   }
 }
+
+/**
+ * INTERNAL API
+ */
+private[akka] class QueueSink[T]() extends GraphStageWithMaterializedValue[SinkShape[T], SinkQueue[T]] {
+  trait RequestElementCallback[E] {
+    val requestElement = new AtomicReference[AnyRef](Nil)
+  }
+
+  type Requested[E] = Promise[Option[T]]
+
+  val in = Inlet[T]("queueSink.in")
+  override val shape: SinkShape[T] = SinkShape.of(in)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    type Received[E] = Try[Option[E]]
+
+    val maxBuffer = module.attributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+    require(maxBuffer > 0, "Buffer size must be greater than 0")
+
+    val buffer = FixedSizeBuffer[Received[T]](maxBuffer + 1)
+    var currentRequest: Option[Requested[T]] = None
+
+    val stageLogic = new GraphStageLogic(shape) with RequestElementCallback[Requested[T]] {
+      override def keepGoingAfterAllPortsClosed = true
+
+      override def preStart(): Unit = {
+        val list = requestElement.getAndSet(callback.invoke _).asInstanceOf[List[Requested[T]]]
+        list.reverse.foreach(callback.invoke)
+        pull(in)
+      }
+
+      private val callback: AsyncCallback[Requested[T]] =
+        getAsyncCallback(promise ⇒ currentRequest match {
+          case Some(_) ⇒
+            promise.failure(new IllegalStateException("You have to wait for previous future to be resolved to send another request"))
+          case None ⇒
+            if (buffer.isEmpty) currentRequest = Some(promise)
+            else sendDownstream(promise)
+        })
+
+      def sendDownstream(promise: Requested[T]): Unit = {
+        val e = buffer.dequeue()
+        promise.complete(e)
+        e match {
+          case Success(_: Some[_]) ⇒ //do nothing
+          case Success(None)       ⇒ completeStage()
+          case Failure(t)          ⇒ failStage(t)
+        }
+      }
+
+      def enqueueAndNotify(requested: Received[T]): Unit = {
+        buffer.enqueue(requested)
+        currentRequest match {
+          case Some(p) ⇒
+            sendDownstream(p)
+            currentRequest = None
+          case None ⇒ //do nothing
+        }
+      }
+
+      setHandler(in, new InHandler {
+        override def onPush(): Unit = {
+          enqueueAndNotify(Success(Some(grab(in))))
+          if (buffer.used < maxBuffer - 1) pull(in)
+        }
+        override def onUpstreamFinish(): Unit = enqueueAndNotify(Success(None))
+        override def onUpstreamFailure(ex: Throwable): Unit = enqueueAndNotify(Failure(ex))
+      })
+    }
+
+    (stageLogic, new SinkQueue[T] {
+      override def pull(): Future[Option[T]] = {
+        val ref = stageLogic.requestElement
+        val p = Promise[Option[T]]
+        ref.get() match {
+          case list: List[_] ⇒
+            if (!ref.compareAndSet(list, p :: list))
+              ref.get().asInstanceOf[Requested[T] ⇒ Unit](p)
```

Comment:
This should be done via a pattern match: if some other thread concurrently tried to pull()—which is forbidden—then the error will be a ClassCastException instead of the more informative IllegalStateException.